ArduCopter/mode_autorotate.cpp

```diff
@@ -26,77 +20,64 @@ bool ModeAutorotate::init(bool ignore_checks)
 
     // Check that mode is enabled, make sure this is the first check as this is the most
     // important thing for users to fix if they are planning to use autorotation mode
-    if (!g2.arot.is_enable()) {
-        gcs().send_text(MAV_SEVERITY_WARNING, "Autorot Mode Not Enabled");
+    if (!g2.arot.enabled()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "AROT: Mode not enabled");
         return false;
     }
 
     // Must be armed to use mode, prevent triggering state machine on the ground
     if (!motors->armed() || copter.ap.land_complete || copter.ap.land_complete_maybe) {
-        gcs().send_text(MAV_SEVERITY_WARNING, "Autorot: Must be Armed and Flying");
+        gcs().send_text(MAV_SEVERITY_WARNING, "AROT: Must be armed and flying");
         return false;
     }
 
-    // Initialise controllers
-    // This must be done before RPM value is fetched
-    g2.arot.init_hs_controller();
-    g2.arot.init_fwd_spd_controller();
-
-    // Retrieve rpm and start rpm sensor health checks
-    _initial_rpm = g2.arot.get_rpm(true);
+    // Initialise controller
+    g2.arot.init();
 
     // Display message 
     gcs().send_text(MAV_SEVERITY_INFO, "Autorotation initiated");
 
-     // Set all intial flags to on
-    _flags.entry_initial = true;
-    _flags.ss_glide_initial = true;
-    _flags.flare_initial = true;
-    _flags.touch_down_initial = true;
-    _flags.landed_initial = true;
-    _flags.level_initial = true;
-    _flags.break_initial = true;
-    _flags.straight_ahead_initial = true;
-    _msg_flags.bad_rpm = true;
+     // Set all initial flags to on
+    _flags.entry_init = false;
+    _flags.glide_init = false;
+    _flags.landed_init = false;
 
     // Setting default starting switches
     phase_switch = Autorotation_Phase::ENTRY;
 
     // Set entry timer
     _entry_time_start_ms = millis();
 
-    // The decay rate to reduce the head speed from the current to the target
-    _hs_decay = ((_initial_rpm/g2.arot.get_hs_set_point()) - HEAD_SPEED_TARGET_RATIO) / AUTOROTATE_ENTRY_TIME;
+    // reset logging timer
+    _last_logged = 0;
 
     return true;
 }
 
-
-
 void ModeAutorotate::run()
 {
     // Current time
     uint32_t now = millis(); //milliseconds
 
+    // Set dt in library
+    float const last_loop_time_s = AP::scheduler().get_last_loop_time_s();
```

Comment:
maybe shift to library?

Reply:
... and perhaps use a last_update_ms and AP_HAL::millis() etc instead?

Reply:
From what I can tell, get_last_loop_time_s() is the correct function.  It does appear to actually measure the time since it was last called in the scheduler.  This is also the same method used in other controllers, such as the position controller.
I have not moved this down into the library.  The Lib does not have an update function that is called every loop.  In the way I have structured it, ModeAutorotate::run() is the only function that is always executed, regardless of flight phase.

Reply:
I understand that this was a blocker for me getting a merge on this.  I just wanted to re-iterate that this method for getting dt is the same as is used in the attitude controller and the position controller:
https://github.com/ArduPilot/ardupilot/blob/781b6bd8f869641240d65065ef6100e97e698327/ArduCopter/Attitude.cpp#L13-L15
So should be absolutely fine for the autorotation case, as we are using the same dt as the lower level attitude controller.

Reply:
given it is a FAST_TASK() I think this is OK